packages/flutter/lib/src/cupertino/context_menu.dart

```diff
@@ -0,0 +1,1099 @@
+  Rect _decoyChildEndRect;
+
+  OverlayEntry _lastOverlayEntry;
+  double _childOpacity = 1.0;
+  _ContextMenuRoute<void> _route;
+
+  @override
+  void initState() {
+    super.initState();
+    _decoyController = AnimationController(
+      duration: const Duration(milliseconds: 600),
+      vsync: this,
+    );
+    _decoyController.addStatusListener(_onDecoyAnimationStatusChange);
+  }
+
+  // Determine the _ContextMenuOrientation based on the location of the original
+  // child in the screen.
+  _ContextMenuOrientation get _contextMenuOrientation {
+    final Rect childRect = _getRect(_childGlobalKey);
+    final double screenWidth = MediaQuery.of(context).size.width;
+
+    final double center = screenWidth / 2;
+    final bool centerDividesChild = childRect.left < center && childRect.right > center;
+    final double minCenterWidth = screenWidth / 4;
+    final double maxCenterWidth = screenWidth / 3;
+    final bool isCenterWidth = childRect.width >= minCenterWidth
+      && childRect.width <= maxCenterWidth;
+    if (centerDividesChild && isCenterWidth) {
+      return _ContextMenuOrientation.center;
+    }
+
+    if (childRect.center.dx > center) {
+      return _ContextMenuOrientation.right;
+    }
+
+    return _ContextMenuOrientation.left;
+  }
+
+  // Push the new route and open the ContextMenu overlay.
+  void _openContextMenu() {
+    HapticFeedback.lightImpact();
+
+    setState(() {
+      _childOpacity = 0.0;
+    });
+
+    _route = _ContextMenuRoute<void>(
+      actions: widget.actions,
+      barrierLabel: 'Dismiss',
+      filter: ui.ImageFilter.blur(
+        sigmaX: 5.0,
+        sigmaY: 5.0,
+      ),
+      onTap: widget.onTap,
+      contextMenuOrientation: _contextMenuOrientation,
+      previousChildRect: _decoyChildEndRect,
+      builder: (BuildContext context) {
+        return widget.preview ?? widget.child;
+      },
+    );
+    Navigator.of(context, rootNavigator: true).push<void>(_route);
+    _route.animation.addStatusListener(_routeAnimationStatusListener);
+  }
+
+  // The OverlayEntry that positions widget.child directly on top of its
+  // original position in this widget.
+  OverlayEntry get _overlayEntry {
+    final Rect childRect = _getRect(_childGlobalKey);
+    _decoyChildEndRect = Rect.fromCenter(
+      center: childRect.center,
+      width: childRect.width * _kOpenScale,
+      height: childRect.height * _kOpenScale,
+    );
+
+    // TODO(justinmc): Using Overlay.of(context) means that this pops on top of
+    // the AppBar, but it doesn't on native. Is there any way around this? The
+    // reason that I do this in an Overlay is that it needs to sit on top of
+    // widgets adjacent to it.
+    // Also, I noticed that I'm having the same theme problem where a widget in
+    // the Overlay might not be themed the same as it is otherwise. Sizing can
+    // also potentially be a problem if the child was sized by its parent before.
+    return OverlayEntry(
```

Comment:
Is it possible to avoid using the additional `OverlayEntry`? So we don't have to manage it and thus avoid the `Future.delayed` hack. Maybe make the expand animation part of the `ContextMenuRoute` push transition, so that `_kOpenScale` will be <= 1.0?

Reply:
I experimented with a few different ways to do this.  For not using an Overlay, it seems to come down to the fact that the child will appear behind its siblings during its bounce animation.  This looked really obvious to me when I tried it.  I updated the comment here to be a bit more clear about the situation at least.